Feature/categoria

fernandolfla opened the pull request.

fernandolfla committed:
Adicionar um migration
Validar categoria com o mesmo nome ativo
Acionar endpoint para buscar todas as categorias ativas
Adicionar um enpoint para inserção
Adicionar um endpoint para desativar

fernandolfla committed:
Busca de inativos

fernandolfla committed:
Adicionar um endpoint para Atualizar
Adicionar um endpoint para ativar

fernandolfla reviewed (commented):
Revisado

FagnerFun commented on Pizzaria_back/Pizzaria_back/Controllers/CategoriaController.cs:
```diff
@@ -0,0 +1,53 @@
+            _service = service;
+        }
+
+        [HttpGet("Buscar")]
```
o que acha de alteramos a rota para [HttpGet("[Action]")], visto que o endpoint tem o mesmo nome do metodo

fernandolfla replied:
Entendi, os 2 gets estavam travando o swegger, então assim deu certo. Alterado.

FagnerFun commented on Pizzaria_back/Pizzaria_back/Controllers/CategoriaController.cs:
```diff
@@ -0,0 +1,53 @@
+
+        [HttpGet("BuscarTudo")]
+        public IActionResult BuscarTudo()
+           => Ok(_service.BuscarTudo().Select(x => (CategoriaResponse)x));
```
Qual a diferença entre o metodo buscar e o buscartudo ?

fernandolfla replied:
Buscar lista apenas as categorias ativas, para montar a tela de produtos, pizza e tal. Buscar tudo lista as categorias ativas e inativas para, na tela de cadastro usar algo como uma switch para mudar esse status.

fernandolfla replied:
Done

FagnerFun commented on Pizzaria_back/Pizzaria_back/Controllers/Model/Response/CategoriaResponse.cs:
```diff
@@ -0,0 +1,23 @@
+        public int Id { get; set; }
+        public string Nome { get; set; }
+        public bool Pizza { get; set; }
+        public bool Ativo { get; set; }
```
precisamos ter o campo ativo aqui ?

fernandolfla replied:
Como teremos o update de status ativo da categoria, imaginei da necessidade de saber no front o ID, para retorno e o Ativo para o usuário, para quando o objeto retornar podermos atualizar com um update e economizar um endpoit "habilitar" .

FagnerFun commented on Pizzaria_back/Pizzaria_back/Interfaces/Repository/IBaseRepository.cs:
```diff
@@ -9,6 +9,6 @@ public interface IBaseRepository<T> where T : class
         List<T> Buscar();
         T? Buscar(int id);
-        
+
```
podemos remover o espaço adicionado ?

fernandolfla replied:
Devemos, removido.

FagnerFun commented on Pizzaria_back/Pizzaria_back/Models/Categoria.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class Categoria : DbEntity
+    {
+        public Categoria() 
```
será que o objeto que categoria que é criado não será construido com ativo falso, visto que não chamamos o construtor da classe pai e não alteramos a propriedade ativo no nosso construtor ?

sugestão : public Categoria() : base(){ ... }

fernandolfla replied:
Em DbEntity tem o construtor de ativo = true. Precisamos acinar ele novamente em todas as classes herdadas?
public class DbEntity
    {
        public DbEntity()
        {
            Ativo = true;
        }
        public int Id { get; set; }
        public bool Ativo { get; set; }
    }

fernandolfla replied:
E também em na configuração da migração temos ativo = true para todo DbEntity.

FagnerFun commented on Pizzaria_back/Pizzaria_back/Models/Produto.cs:
```diff
@@ -7,7 +7,9 @@ public class Produto : DbEntity
-        public bool Fracionado { get; set; }
+
+        public int CategoriaId { get; set; }
+        public Categoria Categoria { get; set; }
```
esse campo categoria é referente ao relacionamento do Entity Framework?, se sim podemos modifica-lo para virtual ?

fernandolfla replied:
Modificado

FagnerFun commented on Pizzaria_back/Pizzaria_back/Models/Tipo.cs:
```diff
@@ -5,7 +5,7 @@ namespace Pizzaria_back.Models
-        public int ProdutoId { get; set;}
-        public Produto Produto;// { get; set; } verificar com fagner a questão do migration
+        public int ProdutoId { get; set; }
+        public Produto Produto { get; set; }
```
podemos adicionar como virtual o campo produto ?

FagnerFun commented on Pizzaria_back/Pizzaria_back/Models/Tipo_Tamanho.cs:
```diff
@@ -3,9 +3,9 @@
     {
         public int TipoId { get; set; }
-        public Tipo Tipo;
+        public Tipo Tipo { get; set; }
```
podemos adicionar como virtual o campo Tipo e Tamanho?

fernandolfla replied:
Feito, todos os campos das entidades relacionadas foram alterados para virtual.

FagnerFun commented on Pizzaria_back/Pizzaria_back/Repository/BaseRepository.cs:
```diff
@@ -46,5 +46,6 @@ public List<T> Buscar()
                 .Where(x => x.Id == id && x.Ativo == true)
                 .FirstOrDefault();
 
+
```
Podemos remover os espaços em brancos ?

fernandolfla replied:
Devemos!

FagnerFun commented on Pizzaria_back/Pizzaria_back/Repository/CategoriaRepository.cs:
```diff
@@ -0,0 +1,26 @@
+        public Categoria? Buscar(string nome)  //categoria não podem ser iguais
+            => _applicationDbContext.Categorias
+                         .FirstOrDefault(x => x.Nome.Equals(nome));
+        public Categoria? Buscar(bool ativo, int id) 
```
podemos remover o parametro ativo, visto que ele não esta sendo utilizado ?

fernandolfla replied:
Eu queria aproveitar o nome do método e sobrecarregar, por isso utilizei um argumento a mais, que quando ele receber um Buscar(true,id) ele use a sobrecarga de dois argumentos para buscar mesmo os que estão desabilitados. Posso sobrepor o método ? Tem alguma alguma sugestão?

fernandolfla replied:
Conclusão, não faz sentido ter uma busca para deletar se categorias não serão deletadas, serão ativadas e desativadas. Removido.

FagnerFun commented on Pizzaria_back/Pizzaria_back/Repository/CategoriaRepository.cs:
```diff
@@ -0,0 +1,26 @@
+            _applicationDbContext = applicationDbContext;
+        }
+
+        public Categoria? Buscar(string nome)  //categoria não podem ser iguais
```
se a ideia deste metodo é apenas validar se existe uma categoria com o mesmo nome, podemos altera-lo para um retorno do tipo booleano, e a utilizar o metodo Any ?, pensando no desempenho do banco, ao inves de trazer uma query com varias colunas ser uma query com uma coluna unica de verdadeiro e falso

Ex: _applicationDbContext.Categorias.Any(x => x.Nome.Equals(nome))

fernandolfla replied:
Certo, alterado.

FagnerFun commented on Pizzaria_back/Pizzaria_back/Repository/Configuration/CategoriaConfiguration.cs:
```diff
@@ -0,0 +1,16 @@
+        {
+            public void Configure(EntityTypeBuilder<Categoria> builder)
+            {
+                builder.Property(x => x.Nome).HasMaxLength(255)
```
Não seria interessante adicionar a configuração das outras colunas da tabela categoria ?

fernandolfla replied:
A entidade possue somente Nome e Pizza, booleano, que será passado. Os atributos herdados possuem propriedade propria, de id = PK  auto e ativo = true. Teria alguma sugestão?

fernandolfla replied:
Pizza foi mudado para padrão false, até por que dois produtos  são  pizza;

fernandolfla committed:
Removido espaços
Redirecionamento com Action em categorias

fernandolfla committed:
Virtual nos campos que são apenas de construção de relação

fernandolfla committed:
Alterado método de busca por categoria repetida para bool any

fernandolfla committed:
Remoção do deletar
Reescalonamento dos endpoints
Buscar e Buscar tudo definido para um único endpoint

fernandolfla commented:
Para revisão novamente.
